File: ui/src/app/lib/api/governance.ts
/**
 * Copyright 2025 The Wardyn Authors
 * SPDX-License-Identifier: Apache-2.0
 */

// Governance profiles (0.7) — the seven securityOps routes behind the
// Governance screen. Mirrors internal/api/governance.go; every route is under
// /api/v1 via wfetch.
//
// The wire types live HERE rather than in lib/types/ because this is a NEW
// domain with exactly one consumer (the Governance screen) — a second module
// holding four interfaces nothing else imports is a file to keep in sync for no
// reader. Move them to lib/types/governance.ts the day a second domain needs
// them.
import type { CapabilitySubjectType, ConfinementClass, RunPolicySpec } from "../types";
import { asJson, errText, HttpError, unwrapList, wfetch } from "./core";

// types.GovernanceLimits. ALL are `omitempty` on the wire, so an unrestricted
// profile arrives with the keys absent — optional here for the same reason,
// and `!!limits.deny_x` is how every read is written.
export interface GovernanceLimits extends RunLimits {
  deny_task_mode_exec?: boolean;
  deny_interactive?: boolean;
  // types.GovernanceLimits.DenyUserDrive (0.7 user drives) — the door the
  // profile editor's third LimitRow writes. A run under this profile mounts no
  // user drive even when one is allocated to the person; denyMemberDrive's 403
  // is what enforces it, and this is only what the editor authors.
  deny_user_drive?: boolean;
  // 0/absent is unlimited (R4/F032). The editor's LimitNumberRow writes this.
  max_concurrent_runs?: number;
  // types.GovernanceLimits.MaxEphemeralDiskMiB (0.7.2) — the ephemeral scratch
  // ceiling. 0/absent is unlimited. A CLAMP, not a refusal: a run asking for
  // more is capped at dispatch and warned, never 403'd.
  max_ephemeral_disk_mib?: number;
  // types.GovernanceLimits.MaxDriveSizeMiB (0.7.2) — the per-principal user
  // drive ceiling, folded with the deployment's own in one min(). 0/absent is
  // unlimited.
  max_drive_size_mib?: number;
  // types.GovernanceLimits.AutonomyRubric (0.8, #77/#99) — maps a run's
  // posture to a permitted autonomy level. Absent means no rubric, same as
  // every field above: the Go side is a pointer so an unrestricted profile
  // still marshals `limits: {}`. Resolution/enforcement land in #97; this
  // mirror exists so the type is in step from the day the field appears.
  autonomy_rubric?: AutonomyRubric;
}

// types.RunLimits (long-holds rev 4 §2.2) — embedded in GovernanceLimits, so
// the seven keys sit flat on `limits`; a run carries the same set, captured at
// create (AgentRun.run_limits). Absent/0/false keeps today's behaviour: no end,
// the deployment's approval expiry as the wait, no idle pause.
export interface RunLimits {
  max_end_ahead_sec?: number;
  default_end_sec?: number;
  allow_no_end?: boolean;
  max_wait_sec?: number;
  default_wait_sec?: number;
  user_changes_limits?: boolean;
  pause_idle_after_sec?: number;
}

// types.AutonomyLevel — L0 (most supervised) through L3 (least). The codes
// stay internal; the console renders plain labels (#93), not spelled here yet.
export type AutonomyLevel = "L0" | "L1" | "L2" | "L3";

// types.AutonomyRubric — nine closed fields, three egress postures, three
// secret postures, three confinement classes, each absent (caps nothing) or
// one of the four levels. See internal/types/governance.go for what each
// posture means.
export interface AutonomyRubric {
  egress_open?: AutonomyLevel;
  egress_reviewed?: AutonomyLevel;
  egress_sealed?: AutonomyLevel;
  secrets_powerful?: AutonomyLevel;
  secrets_baseline?: AutonomyLevel;
  secrets_none?: AutonomyLevel;
  confinement_cc1?: AutonomyLevel;
  confinement_cc2?: AutonomyLevel;
  confinement_cc3?: AutonomyLevel;
}

// One of AutonomyRubric's nine own field names — what internal/composer/
// autonomy.go's applicableAutonomyCaps names a cap by, and what
// AutonomyResolution.BoundBy (below) carries. keyof, not a hand-typed union,
// so the two can never drift apart.
export type AutonomyRubricRowKey = keyof AutonomyRubric;

// applicableAutonomyCaps' own fixed field order (internal/composer/
// autonomy.go) — egress, then secrets, then confinement — the order
// FoldAutonomy lists tied causes in, so a rendered bound_by sentence reads the
// same on Review and at launch. Also the row order profile-rubric.tsx draws
// the editor's three groups in.
export const AUTONOMY_RUBRIC_ROW_KEYS: AutonomyRubricRowKey[] = [
  "egress_open",
  "egress_reviewed",
  "egress_sealed",
  "secrets_powerful",
  "secrets_baseline",
  "secrets_none",
  "confinement_cc1",
  "confinement_cc2",
  "confinement_cc3",
];

// AutonomyLevel's own weakest -> strongest ladder (internal/types/
// governance.go's AutonomyLevel.Rank()) — mirrors lib/types/runs.ts's
// CC_ORDER for the same reason: lib/ may not import from components/, so the
// rank order lives here beside the type it orders, and the display labels
// (AUTONOMY_META) live in components/wardyn/autonomy-meta.ts instead.
export const AUTONOMY_LEVEL_ORDER: AutonomyLevel[] = ["L0", "L1", "L2", "L3"];

// types.AutonomyEgressPosture / types.AutonomySecretsPosture — the two graded
// halves of a run's three-axis posture (internal/composer/autonomy.go).
export type AutonomyEgressPosture = "open" | "reviewed" | "sealed";
export type AutonomySecretsPosture = "powerful" | "baseline" | "none";

// types.AutonomyPosture — the three-axis shape #97's resolveRunAutonomy folds
// against a profile's AutonomyRubric.
export interface AutonomyPosture {
  egress: AutonomyEgressPosture;
  secrets: AutonomySecretsPosture;
  confinement: ConfinementClass;
}

// types.AutonomyResolution (0.8 #97/#93) — what resolveRunAutonomy decided for
// one run: the level, the posture that produced it, and every rubric row that
// bound the result.
//
// bound_by IS A LIST, not a string — a wire decision, not a rendering choice
// (#96's ruling on #93). The fold is a min() over three axes, so rows TIE at
// the resolved level routinely; naming only the first would send an admin to
// raise a row the level would not actually move on. governance-copy.ts's
// autonomyBoundSentence composes the FULL list into one sentence, never just
// bound_by[0].
export interface AutonomyResolution {
  level: AutonomyLevel;
  posture: AutonomyPosture;
  bound_by?: AutonomyRubricRowKey[];
}

// types.GovernanceProfile — one named, assignable ceiling.
export interface GovernanceProfile {
  id: string;
  name: string;
  ceiling: RunPolicySpec;
  limits: GovernanceLimits;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

// types.GovernanceAssignment — one subject bound to one profile.
export interface GovernanceAssignment {
  id: string;
  subject_type: CapabilitySubjectType;
  subject: string;
  profile_id: string;
  priority: number;
  created_at: string;
  created_by?: string;
}

// GET /governance's body: the whole picture in one read.
export interface GovernanceSnapshot {
  profiles: GovernanceProfile[];
  assignments: GovernanceAssignment[];
}

// The POST/PUT body. id/created_at/updated_at/created_by are never sent — the
// server assigns provenance (governanceProfileRequest).
export interface GovernanceProfileInput {
  name: string;
  ceiling: RunPolicySpec;
  limits: GovernanceLimits;
}

// governanceProfileResponse: the saved row plus the OMISSION warnings the
// server composed. `warnings` is `omitempty` on the wire, so it is coerced to
// an array here and the screen never guards it.
//
// The console renders those strings VERBATIM, in response order — they are the
// server's own prose, more specific than any frozen sentence could be
// (governance-prompt.md §7.4).
export interface GovernanceProfileWrite {
  profile: GovernanceProfile;
  warnings: string[];
}

// The POST /governance/assignments body.
export interface GovernanceAssignmentInput {
  subject_type: CapabilitySubjectType;
  subject: string;
  profile_id: string;
  priority: number;
}

// The POST /governance/preview body — the two claim lists
// ResolveGovernanceProfile itself takes, so the server has nothing to derive.
//
// The console's claims field is kind-LESS (one textarea, the People step's own
// shape), so it sends every typed line in BOTH lists and the server offers each
// to both tiers. That is exactly what POST /access/preview already receives
// ({roles: lines, groups: lines}), and it is honest because the ANSWER names
// the tier that matched.
export interface GovernancePreviewInput {
  user_subjects: string[];
  groups: string[];
}

// One kind-LESS claims textarea onto the two typed wire lists both previews
// take — the governance one and the user-drive one, which post the identical
// body and are folded by the identical server-side normalizer.
//
// The ONE thing decided here is ORDER, never precedence. The user tier's
// tie-break is POSITION in user_subjects (the resolver's array_position) and
// the enforcement path builds it sign-in-subject first, email second
// (capabilitySubjects) — so an "@" line sorts last and a preview agrees with
// what actually binds the member. The drive resolver reads the same order
// positionally to pick which claim an `email_local` home is named from. Array
// .prototype.sort is stable, so every other line keeps the order it was typed
// in. The ranking itself stays in SQL.
export function previewClaims(claims: string): GovernancePreviewInput {
  const lines = claims
    .split("\n")
    .map((c) => c.trim())
    .filter(Boolean);
  return {
    user_subjects: [...lines].sort((a, b) => Number(a.includes("@")) - Number(b.includes("@"))),
    groups: lines,
  };
}

// governancePreviewResponse: which profile would bind a principal carrying
// those claims, and the tier of the assignment that won.
//
// EVERY field is `omitempty` on the wire, so an empty object is the answer for
// "no assignment matched" — the deployment ceiling. Same absent-key doctrine as
// DefaultPolicy's governance_profile_name; the two present fields ship
// together, so the screen tests `profile_name` and reads the tier beside it.
export interface GovernancePreview {
  profile_id?: string;
  profile_name?: string;
  matched_tier?: CapabilitySubjectType;
}

// The server composes a grant-bound refusal as "invalid ceiling: " + the
// comparator's own error, and EVERY leg of that comparator opens "eligible
// grant " (governanceGrantWithinCeiling, internal/api/governance_grantbound.go).
// A spec-VALIDATION 400 shares the "invalid ceiling: " prefix but never that
// clause, which is what makes this a safe discriminator: the console heads the
// grant-bound body with GRANT_BOUND_TITLE and renders every other 400 as the
// server's own message under no heading at all.
//
// A discriminator, not copy: it is matched against the wire, never rendered.
const GRANT_BOUND_PREFIX = "invalid ceiling: eligible grant ";

export function isGrantBoundError(e: unknown): boolean {
  return e instanceof HttpError && e.status === 400 && e.message.startsWith(GRANT_BOUND_PREFIX);
}

export const governance = {
  // GET /api/v1/governance -> {profiles, assignments}. Nil Go slices encode as
  // null, so both halves are coerced and every caller can map over them.
  async getGovernance(): Promise<GovernanceSnapshot> {
    const res = await wfetch("/governance", { method: "GET" });
    const body = await asJson<Partial<GovernanceSnapshot>>(res);
    return {
      profiles: unwrapList<GovernanceProfile>(body.profiles),
      assignments: unwrapList<GovernanceAssignment>(body.assignments),
    };
  },

  // POST /api/v1/governance/profiles -> 201 {profile, warnings}. A 400 whose
  // message opens "invalid ceiling: eligible grant " is the monotone-⊆ bound
  // refusing a profile that would MINT credential eligibility; every 4xx
  // surfaces as an HttpError carrying the server's own message.
  async createProfile(input: GovernanceProfileInput): Promise<GovernanceProfileWrite> {
    return writeProfile(await wfetch("/governance/profiles", { method: "POST", body: JSON.stringify(input) }));
  },

  // PUT /api/v1/governance/profiles/{id} -> 200 {profile, warnings}. Rename
  // included: ON DELETE RESTRICT makes delete-and-recreate impossible for an
  // assigned profile, so the update path has to carry the name.
  async updateProfile(id: string, input: GovernanceProfileInput): Promise<GovernanceProfileWrite> {
    return writeProfile(
      await wfetch(`/governance/profiles/${encodeURIComponent(id)}`, { method: "PUT", body: JSON.stringify(input) }),
    );
  },

  // DELETE /api/v1/governance/profiles/{id} -> 204.
  //
  // The 409 MUST reach the caller: it is the ON DELETE RESTRICT refusal, and it
  // is authoritative for the race the console's own assignment count cannot see
  // (another admin assigning the profile a second ago). A 404 is tolerated as
  // "already gone" — the row is absent either way, which is what was asked for.
  async deleteProfile(id: string): Promise<void> {
    const res = await wfetch(`/governance/profiles/${encodeURIComponent(id)}`, { method: "DELETE" });
    if (!res.ok && res.status !== 404) {
      throw new HttpError(res.status, await errText(res));
    }
  },

  // POST /api/v1/governance/assignments -> the stored row. Keyed on the natural
  // (subject_type, subject): re-assigning a subject REPOINTS its single row
  // (200) rather than accumulating a second (201). The console reloads either
  // way, so the status is not surfaced.
  async upsertAssignment(input: GovernanceAssignmentInput): Promise<GovernanceAssignment> {
    const res = await wfetch("/governance/assignments", { method: "POST", body: JSON.stringify(input) });
    return asJson<GovernanceAssignment>(res);
  },

  // DELETE /api/v1/governance/assignments/{id} -> 204. This is the SUPPORTED
  // way to widen a principal back to the deployment ceiling.
  async deleteAssignment(id: string): Promise<void> {
    const res = await wfetch(`/governance/assignments/${encodeURIComponent(id)}`, { method: "DELETE" });
    if (!res.ok && res.status !== 404) {
      throw new HttpError(res.status, await errText(res));
    }
  },

  // POST /api/v1/governance/preview -> 200, the resolved profile or {}.
  //
  // The dry run runs THE resolver server-side —
  // Store.ResolveGovernanceProfile, the same call the enforcement path makes on
  // every run. There is deliberately NO client-side precedence here: a second
  // implementation of the SQL ORDER BY precedence rule in TypeScript is a
  // second implementation of the answer. It saves nothing and enforces
  // nothing, but it must not be able to disagree with what actually binds a
  // member.
  async previewGovernance(input: GovernancePreviewInput): Promise<GovernancePreview> {
    const res = await wfetch("/governance/preview", { method: "POST", body: JSON.stringify(input) });
    return asJson<GovernancePreview>(res);
  },
};

async function writeProfile(res: Response): Promise<GovernanceProfileWrite> {
  const body = await asJson<GovernanceProfileWrite>(res);
  return { profile: body.profile, warnings: unwrapList<string>(body.warnings) };
}

File: internal/types/governance.go
// Copyright 2025 The Wardyn Authors
// SPDX-License-Identifier: Apache-2.0

// Governance profiles (migration 0052): the wire + store types for an
// ASSIGNABLE ceiling and the row that binds one to a subject.
//
// Named `governance*` throughout, deliberately: "profile" already means two
// unrelated things in this tree — the Recording-Mode synthesis result and the
// workspace scan profile — and a third would make every grep ambiguous at
// exactly the surface where a mistaken read is a containment bug.
package types

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// GovernanceLimits carries the autonomy switches that BOUND A REQUEST rather
// than a policy, which is why they live here and not in RunPolicySpec.
//
// The distinction is load-bearing. RunPolicySpec describes what a sandbox may
// reach once it is running, and every field of it is enforced OUTSIDE the
// sandbox by the proxy. The booleans instead refuse a run SHAPE before it
// exists, because each names a way to route AROUND the tool gate entirely, or
// (DenyUserDrive) a way for a run to leave state behind it:
//
//   - DenyTaskModeExec: task_mode=exec runs a bare command with no agent and no
//     toolgate in the loop, so no tool_rules ceiling can bind it. A profile that
//     wants supervised tool use has to be able to say "not through that door".
//   - DenyInteractive: an interactive run REFUSES tool_approvals=hold by design
//     (a human at the attach pane is the supervision), so a profile cannot
//     express "supervised" through tool_rules on that lane either. This is the
//     lever built for it.
//   - DenyUserDrive: a user drive is a tree that OUTLIVES the run, so a member
//     who may mount one can persist anything the sandbox produced past the
//     sandbox's own lifetime. No tool_rules ceiling describes that, because the
//     escape is the storage, not the tool.
//   - MaxConcurrentRuns is the odd one out — a QUOTA, not a door. It bounds how
//     many runs one member holds at once rather than what any single run may be,
//     which is why its enforcement site answers 422 with no authz.denied while
//     the booleans answer 403 with one.
//   - MaxEphemeralDiskMiB and MaxDriveSizeMiB are the two SIZES (0.7.2), and
//     they are neither doors nor quotas: they CLAMP. A run or a drive at the
//     bound is capped and told so, never refused — `disk_mib` is authored on
//     policies, so a 422 would break every stored policy the day a limit is
//     first written. Their enforcement sites are the two places that hold both
//     the profile and the deployment ceiling: dispatch for the scratch size,
//     newResolvedDrive for the drive.
//
// A CLOSED struct with `omitempty` on every field, not a map: the set is small,
// complete, and validated by the Go type itself, so migration 0052 puts no
// CHECK on the limits column at all (the 0042 doctrine — one closed Go
// definition, validated at the write boundary, zero DDL for the next member).
// EVERY zero value means "unrestricted", so a profile that omits limits behaves
// exactly as one written before this struct had fields — the absent-row
// back-compat rule applied one level down.
type GovernanceLimits struct {
	// DenyTaskModeExec refuses task_mode=exec for a member under this profile.
	DenyTaskModeExec bool `json:"deny_task_mode_exec,omitempty"`
	// DenyInteractive refuses an interactive run for a member under this
	// profile. Evaluated against POST-COERCION interactivity by its enforcement
	// site: a request that simply OMITS the task coerces to interactive later in
	// validation, so a raw "did the caller ask for interactive" read is evaded
	// by leaving a field out.
	DenyInteractive bool `json:"deny_interactive,omitempty"`
	// MaxConcurrentRuns caps how many NON-TERMINAL runs a member under this
	// profile may hold at once. 0 is unlimited — the same zero-value rule the
	// two booleans follow, so limits authored before this field existed keep
	// meaning what they meant.
	MaxConcurrentRuns int `json:"max_concurrent_runs,omitempty"`
	// DenyUserDrive refuses a USER DRIVE mount for a member under this profile:
	// their run may not carry drive.enabled at all, whatever an admin has
	// allocated them.
	//
	// A DOOR, NOT A QUOTA, which is why it is a bool beside the other two
	// rather than a size beside MaxConcurrentRuns. A drive is a writable tree
	// that OUTLIVES the run — the one piece of state an agent can leave behind
	// — so "how big" is the wrong question for a ceiling to ask; "may this
	// principal persist anything at all" is the right one, and it is the same
	// shape as the two refusals above (403 with an authz.denied row, not a 422
	// quota answer).
	//
	// Zero means unrestricted, like every other field here: a profile written
	// before drives existed keeps meaning exactly what it meant, and a
	// deployment that never allocates a drive is unaffected either way.
	DenyUserDrive bool `json:"deny_user_drive,omitempty"`
	// MaxEphemeralDiskMiB caps the EPHEMERAL scratch a member's run under this
	// profile may be given — the writable layer a sandbox gets when it mounts no
	// drive. 0 is unlimited, the same zero-value rule every field here follows.
	//
	// A CLAMP, NOT A DOOR, which is why it is a size beside MaxConcurrentRuns
	// rather than a bool beside the three refusals. `disk_mib` is authored on
	// POLICIES, so refusing a run that asks for more would break every stored
	// policy the day an admin first writes a limit; the run is capped and told
	// so, in composer.Clamp's own idiom ("resources capped to operator
	// maximum"). An authorized caller at a bound therefore earns no
	// authz.denied row — nothing was denied.
	//
	// ENFORCED AT DISPATCH, in ONE place (runs_dispatch.go, beside the ceiling
	// deny re-assertion), folded together with the deployment's own
	// storage.ephemeral.max_disk_mib ceiling — never on the create path, whose
	// resourceLimitsToRunner is a pure mapper with neither the ceiling nor the
	// site config in scope. Assigned members only; operators are exempt.
	//
	// WHETHER THE CAP BINDS depends on the substrate: it is a request the runner
	// makes of Kubernetes or Docker, and Docker's overlay2 does not enforce a
	// size at all. Render it through StorageEnforcement and never claim a cap
	// the substrate does not keep.
	MaxEphemeralDiskMiB int `json:"max_ephemeral_disk_mib,omitempty"`
	// MaxDriveSizeMiB caps how large a USER DRIVE may be for a member under this
	// profile. 0 is unlimited. The governance twin of DenyUserDrive: that field
	// answers "may this principal persist anything at all", this one answers
	// "how much" — and a profile can carry either without the other.
	//
	// PER PRINCIPAL, and that is an open question the shape argument has to
	// name: a drive today is one tree belonging to one subject, so a ceiling on
	// its size is a ceiling on that person. Team-shared drives (0.8) add a scope
	// axis on the same row rather than a second field — a shared drive's ceiling
	// is not the sum of its members' and must not be derived from one. Until
	// then, "per principal" is the whole meaning.
	//
	// CLAMPED IN newResolvedDrive (user_drives_resolve.go), the one scope that
	// holds BOTH facts — never at grant write, where the profile binding a
	// subject is claims-resolved and unreadable from the row. It folds with the
	// deployment's own storage.user_drive.max_size_mib in one min() expression,
	// so launch, /me and POST /drives/preview cannot disagree about a drive's
	// size.
	MaxDriveSizeMiB int `json:"max_drive_size_mib,omitempty"`
	// AutonomyRubric maps a run's posture to a permitted AutonomyLevel for a
	// member under this profile (0.8, #77). A POINTER: `omitempty` never omits a
	// struct value, so a plain (non-pointer) field would put
	// `"autonomy_rubric":{}` on every profile's wire body, including one
	// authored before this field existed — TestGovernanceLimitsWireRoundTrip
	// pins that an unrestricted profile still marshals `limits: {}` byte for
	// byte, the same zero-value rule every field above follows. Nil means "no
	// rubric": resolveRunAutonomy (#97) treats it exactly like a member with no
	// assigned profile at all.
	AutonomyRubric *AutonomyRubric `json:"autonomy_rubric,omitempty"`
}

// AutonomyLevel is one rung on the autonomy ladder a governance profile's
// AutonomyRubric caps against, L0 (most supervised) through L3 (least). The
// codes stay internal — the console renders plain labels — the same way
// ConfinementClass's CC1/CC2/CC3 do (0.8 #77):
//
//   - AutonomyL0 "attended": interactive only, supervised seeding.
//   - AutonomyL1 "gated": adds non-interactive runs, but tool approvals are
//     derived to `hold`.
//   - AutonomyL2 "unattended": adds auto-approval and seeded auto tools.
//   - AutonomyL3: adds `task_mode=exec`, the door that routes around every
//     other gate, so it is the top rung.
type AutonomyLevel string

const (
	AutonomyL0 AutonomyLevel = "L0"
	AutonomyL1 AutonomyLevel = "L1"
	AutonomyL2 AutonomyLevel = "L2"
	AutonomyL3 AutonomyLevel = "L3"
)

// Valid reports whether l is one of the four defined rungs. Unlike
// ConfinementClass (which has no such gate — every caller of Rank already
// tolerates rank 0), AutonomyRubric needs one: an author-facing field, so a
// typo must 400 rather than silently rank as "below L0".
func (l AutonomyLevel) Valid() bool {
	switch l {
	case AutonomyL0, AutonomyL1, AutonomyL2, AutonomyL3:
		return true
	}
	return false
}

// Rank orders AutonomyLevel weakest (most supervised) -> strongest (least),
// mirroring ConfinementClass.Rank(): resolving a rubric folds several
// applicable caps to their MINIMUM level (#97), and Rank is what "minimum"
// compares on. An unrecognised value ranks below L0 so it never wins a min()
// against a real level.
func (l AutonomyLevel) Rank() int {
	switch l {
	case AutonomyL0:
		return 0
	case AutonomyL1:
		return 1
	case AutonomyL2:
		return 2
	case AutonomyL3:
		return 3
	default:
		return -1
	}
}

// AutonomyRubric maps a run's posture to a permitted AutonomyLevel. Nine
// closed fields — three egress postures, three secret postures, three
// confinement classes — each unset (that posture caps nothing) or one of the
// four levels (0.8 #77's design: "nine closed fields ... each unset or a
// level"). The level a run resolves to is the MINIMUM over every field whose
// posture applies (internal/composer/autonomy.go, #97); an all-unset rubric
// caps nothing, identically to a nil rubric.
//
// A closed struct with `omitempty` on every field, not a map — the same
// GovernanceLimits doctrine this type lives inside of: the set is small,
// complete, and validated by the Go type itself (Validate), so no DDL CHECK
// backs the stored JSON column at all.
type AutonomyRubric struct {
	// EgressOpen caps the level when the run's egress is OPEN: allow-all, or
	// any allowlisted host beyond baseline.
	EgressOpen AutonomyLevel `json:"egress_open,omitempty"`
	// EgressReviewed caps the level when egress is REVIEWED: first-use approval
	// raises approvals, but nothing is wide open.
	EgressReviewed AutonomyLevel `json:"egress_reviewed,omitempty"`
	// EgressSealed caps the level when egress is SEALED: neither of the above.
	EgressSealed AutonomyLevel `json:"egress_sealed,omitempty"`
	// SecretsPowerful caps the level when the run holds a POWERFUL secret: any
	// write-capable grant, an api_key to a non-baseline host, or a
	// git_pat/ssh_key/env_secret grant.
	SecretsPowerful AutonomyLevel `json:"secrets_powerful,omitempty"`
	// SecretsBaseline caps the level when the run holds any grant, none of them
	// powerful.
	SecretsBaseline AutonomyLevel `json:"secrets_baseline,omitempty"`
	// SecretsNone caps the level when the run holds no grant at all.
	SecretsNone AutonomyLevel `json:"secrets_none,omitempty"`
	// ConfinementCC1/CC2/CC3 cap the level by the run's ENFORCED confinement
	// class (an empty enforced class reads as CC1 — see AutonomyPosture).
	ConfinementCC1 AutonomyLevel `json:"confinement_cc1,omitempty"`
	ConfinementCC2 AutonomyLevel `json:"confinement_cc2,omitempty"`
	ConfinementCC3 AutonomyLevel `json:"confinement_cc3,omitempty"`
}

// Validate reports the first field carrying a value that is not a defined
// AutonomyLevel, NAMING that field — governanceLimitsRefusal
// (internal/api/governance.go) prefixes the field name onto its "limits."
// 400 so an admin is told which of the nine to fix, not just "invalid". An
// empty field is always valid: unset means "this posture caps nothing".
func (a AutonomyRubric) Validate() error {
	for _, f := range []struct {
		field string
		level AutonomyLevel
	}{
		{"egress_open", a.EgressOpen},
		{"egress_reviewed", a.EgressReviewed},
		{"egress_sealed", a.EgressSealed},
		{"secrets_powerful", a.SecretsPowerful},
		{"secrets_baseline", a.SecretsBaseline},
		{"secrets_none", a.SecretsNone},
		{"confinement_cc1", a.ConfinementCC1},
		{"confinement_cc2", a.ConfinementCC2},
		{"confinement_cc3", a.ConfinementCC3},
	} {
		if f.level != "" && !f.level.Valid() {
			return fmt.Errorf("%s: %q is not a valid autonomy level", f.field, f.level)
		}
	}
	return nil
}

// AutonomyEgressPosture is one of the three egress states an AutonomyRubric
// caps against (internal/composer/autonomy.go's posture arithmetic, #97).
type AutonomyEgressPosture string

const (
	AutonomyEgressOpen     AutonomyEgressPosture = "open"
	AutonomyEgressReviewed AutonomyEgressPosture = "reviewed"
	AutonomyEgressSealed   AutonomyEgressPosture = "sealed"
)

// AutonomySecretsPosture is one of the three secret-power states an
// AutonomyRubric caps against.
type AutonomySecretsPosture string

const (
	AutonomySecretsPowerful AutonomySecretsPosture = "powerful"
	AutonomySecretsBaseline AutonomySecretsPosture = "baseline"
	AutonomySecretsNone     AutonomySecretsPosture = "none"
)

// AutonomyPosture is a run's three-axis shape — egress reach, secret power,
// enforced confinement class — the input #97's resolveRunAutonomy folds
// against a profile's AutonomyRubric to pick an AutonomyLevel. Computed, never
// stored on its own; it travels inside AutonomyResolution.
type AutonomyPosture struct {
	Egress      AutonomyEgressPosture  `json:"egress"`
	Secrets     AutonomySecretsPosture `json:"secrets"`
	Confinement ConfinementClass       `json:"confinement"`
}

// AutonomyResolution is what resolveRunAutonomy (#97) decides for one run: the
// level, the posture that produced it, and which rubric field bound the
// result — the "level, the posture and what bound it" #77 asks to be
// provenance on the create audit row, the frozen AgentRun.AutonomyLevel, and
// the preflight response. Bound is empty when nothing capped the level (no
// profile, no rubric, or a posture the rubric left unset) — the zero value
// throughout, matching every other GovernanceLimits field's "empty means
// unrestricted" rule.
type AutonomyResolution struct {
	Level   AutonomyLevel   `json:"level"`
	Posture AutonomyPosture `json:"posture"`
	Bound   string          `json:"bound,omitempty"`
}

// GovernanceProfile is one named, assignable ceiling (migration 0052's
// governance_profiles row).
//
// Ceiling is a full RunPolicySpec and is REPLACEMENT semantics, not composition:
// an assigned profile IS the principal's ceiling, and a principal with no
// assignment falls through to the deployment's Config.DefaultPolicy byte for
// byte. Composing the two would mean folding them through composer.Clamp, which
// is NOT a lattice meet (it drops workspace_mounts unconditionally, is
// order-dependent through llm_inspection, and defaults unnamed tools to hold) —
// so a "composed" ceiling would silently lose fields and could not express an
// autonomous profile at all.
//
// Name is the UNIQUE human handle: what an admin assigns by, what the console
// lists, and the last tie-break in the resolver's ORDER BY (so LIMIT 1 is
// deterministic even when priority ties).
type GovernanceProfile struct {
	ID        uuid.UUID        `json:"id"`
	Name      string           `json:"name"`
	Ceiling   RunPolicySpec    `json:"ceiling"`
	Limits    GovernanceLimits `json:"limits"`
	CreatedAt time.Time        `json:"created_at"`
	UpdatedAt time.Time        `json:"updated_at"`
	CreatedBy string           `json:"created_by,omitempty"`
}

// GovernanceAssignment binds one profile to one subject (migration 0052's
// governance_assignments row).
//
// SubjectType REUSES CapabilitySubjectType — the same user/group/all vocabulary
// capability_grants is written against, and the same one capabilitySubjects
// resolves a caller into. A second enum meaning the same three things is the
// dual-matcher drift this codebase already warns about elsewhere: two
// definitions of "who" that disagree by one case is how a deny stops biting.
//
// Priority breaks ties WITHIN a tier (higher wins) — the group tier is where it
// earns its keep, since a member is typically in several groups at once and the
// admin needs to say which group's profile is the operative one. It does NOT
// cross tiers: a user-tier row beats every group-tier row at any priority,
// because an assignment is one admin explicitly naming one principal.
type GovernanceAssignment struct {
	ID          uuid.UUID             `json:"id"`
	SubjectType CapabilitySubjectType `json:"subject_type"`
	Subject     string                `json:"subject"`
	ProfileID   uuid.UUID             `json:"profile_id"`
	Priority    int                   `json:"priority"`
	CreatedAt   time.Time             `json:"created_at"`
	CreatedBy   string                `json:"created_by,omitempty"`
}
